src/coreclr/src/jit/morph.cpp

```diff
@@ -2705,6 +2705,8 @@ void Compiler::fgInitArgInfo(GenTreeCall* call)
         size_t   addrValue           = (size_t)call->gtEntryPoint.addr;
         GenTree* indirectCellAddress = gtNewIconHandleNode(addrValue, GTF_ICON_FTN_ADDR);
         indirectCellAddress->SetRegNum(REG_R2R_INDIRECT_PARAM);
+        // Don't attempt to CSE this constant
+        indirectCellAddress->SetDoNotCSE();
```

Comment:
It is interesting that we are not CSEing indirect cell addresses.

We can perhaps CSE the following code pattern:
```asm
        ...
        9000000B          adrp    x11, [RELOC #0x1de756537c0]
        9100016B          add     x11, x11, #0
        F9400160          ldr     x0, [x11]
        D63F0000          blr     x0
        ...
        9000000B          adrp    x11, [RELOC #0x1de756537c0]
        9100016B          add     x11, x11, #0
        F9400160          ldr     x0, [x11]
        D63F0000          blr     x0
        AA0003EF          mov     x15, x0
       ...
```

```asm
        ...
        9000000B          adrp    x11, [RELOC #0x1de756537c0]
        9100016B          add     x11, x11, #0
        F9400160          ldr     x0, [x11]
                          mov   xR, x0     ; store x0 in some register xR
        D63F0000          blr     x0
        ...
                          mov   x0, xR     ; retrieve xR into x0
        D63F0000          blr     x0
        AA0003EF          mov     x15, x0
       ...
```

With this, we can get an improvement of 8 bytes + 1 elimination of memory access. 
I wrote an analyzer asm to find out how many addresses are CSE candidates and the number is huge. From what I noticed, it would by little over 2MB of size reduction.

`Processed 191816 methods. Found 29246 methods containing 259123 groups.`

Details: [cse-candidates.txt](https://github.com/dotnet/runtime/files/4853383/cse-candidates.txt)

@BruceForstall , @CarolEidt , @AndyAyersMS

Reply:
AnalyzeAsm code that I used to find out CSE candidates: https://github.com/dotnet/jitutils/pull/273

Reply:
I added this limitation because it was causing an assert failure in LSRA.
I will re-check this to see if it is still happens and add a comment if so.

Reply:
In our IR the indirection of r11 is not made explicit until lowering so we can't create a CSE for these indirections at this time:

```
        F9400160          ldr     x0, [x11]
        D63F0000          blr     x0
```

Reply:
Actually the indirection of r11 is never materialized in our current IR:

```
Generating: N011 (  2,  8) [000039] -------N----        t39 =    CNS_INT(h) long   0xd1ffab1e ftn REG x11 $140
IN0002:             adrp    x11, [RELOC #0xd1ffab1e]
IN0003:             add     x11, x11, #0
                                                              /--*  t39    long   
Generating: N013 (???,???) [000144] ------------       t144 = *  PUTARG_REG long   REG x11
                                                              /--*  t144   long   arg0 in x11
Generating: N015 ( 16, 11) [000005] --C---------         t5 = *  CALL help r2r_ind ref    HELPER.CORINFO_HELP_READYTORUN_NEW REG x0 $1c0
IN0004:             ldr     x0, [x11]
Call: GCvars=0000000000000000 {}, gcrefRegs=180000 {x19 x20}, byrefRegs=0000 {}
IN0005:             blr     x0
```

Reply:
If CSE of Constants are enable for arm32 we will hit an assert when we CSE this address:
(So I will add an #ifdef to prevent this on Arm32)

Assertion failed 'candidates != candidateBit' in 'System.Array:ConvertAll(System.__Canon[],System.Converter`2[__Canon,__Canon]):System.__Canon[]' during 'LSRA build intervals' (IL size 64)

Reply:
Sure. Is there a follow-up issue to investigate this ARM failure?